File: arduino_bot_ws/src/arduino_bot_test/scripts/simple_action_server.py
#!/usr/env/bin python3

import rospy
import actionlib
from arduino_bot_test.msg import FibonacciAction, FibonacciResult, FibonacciFeedback


class FibonacciActionServer(object):
    _feedback = FibonacciFeedback()
    _result = FibonacciResult()
    
    def __init__(self, name):
        self._action_name = name
        self._as = actionlib.SimpleActionServer(self._action_name, FibonacciAction, execute_cb=self.execute_callback, auto_start=False)
        self._as.start()
        rospy.loginfo('Simple Action Server Started')
    
    def execute_callback(self, goal, ):
        rospy.loginfo('Goal Received %s' % goal.goal)
        r = rospy.Rate(1)
        success = True    
        self._feedback.sequence = []
        self._feedback.sequence.append(1)
        self._feedback.sequence.append(1)

        for i in range(1, goal.goal - 1):
            if self._as.is_preempt_requested():
                rospy.loginfo('%s Preempted' % self._action_name)
                self._as.set_preempted()
                success = False
                break
             
            self._feedback.sequence.append(self._feedback.sequence[i] + self._feedback.sequence[i-1])
            self._as.publish_feedback(self._feedback)
            r.sleep()

        if success:
            self._result.sequence = self._feedback.sequence
            rospy.loginfo('%s Succeeded' % self._action_name)
            self._as.set_succeeded(self._result)

if __name__ == '__main__':
    rospy.init_node('fibonacci')
    server = FibonacciActionServer('fibonacci')
    rospy.spin()

File: arduino_bot_ws/src/arduino_bot_test/src/simple_action_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include "arduino_bot_test/FibonacciAction.h"

class FibonacciActionServer 
{
public:
    ros::NodeHandle _nh;
    actionlib::SimpleActionServer<arduino_bot_test::FibonacciAction> _as;
    std::string _action_name;
    arduino_bot_test::FibonacciFeedback _feedback;
    arduino_bot_test::FibonacciResult _result;

    FibonacciActionServer(std::string name)
        : _as(_nh, name, boost::bind(&FibonacciActionServer::ExecuteCallback, this, _1), false),
          _action_name(name)
    {
        _as.start();
        ROS_INFO("Simple action server started");

    }

    void ExecuteCallback(const arduino_bot_test::FibonacciGoalConstPtr &goal) 
    {
        ROS_INFO("Goal Received %i", goal->goal);
        ros::Rate rate(1);
        bool success = true;
        
        _feedback.sequence.clear();
        _feedback.sequence.push_back(1);
        _feedback.sequence.push_back(1);

        for( int i = 1; i < goal->goal; i ++)
        {
            if(_as.isPreemptRequested() || !ros::ok())
            {
                ROS_INFO("%s is Preempted", _action_name.c_str());
                _as.setPreempted();
                success = false;
                break;
            }

            _feedback.sequence.push_back(_feedback.sequence[i] + _feedback.sequence[i-1]);
            _as.publishFeedback(_feedback);
            rate.sleep();
        }

        if(success) 
        {
            _result.sequence = _feedback.sequence;
            ROS_INFO("%s is succeeded", _action_name.c_str());
            _as.setSucceeded(_result);
        }
    }


    virtual ~FibonacciActionServer() 
    {

    }


};

int main(int argc, char** argv) 
{
    ros::init(argc, argv, "fibonacci");
    FibonacciActionServer fibonacci_srv("fibonacci");
    
    ros::spin();

    return 0;
}






